File: src/2025-02-23-deepclaude-try/javascript/exercises/practice/variable-length-quantity/variable-length-quantity.js
//
// This is only a SKELETON file for the 'Variable Length Quantity' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const encode = (numbers) => {
  if (typeof number !== 'number' || !Number.isInteger(number) || number < 0 || number > 0xFFFFFFFF) {
    throw new Error('Invalid number');
  }

  if (number === 0) {
    return [0x00];
  }

  let chunks = [];
  while (number > 0) {
    chunks.push(number & 0x7F);
    number = Math.floor(number / 128);
  }
  chunks = chunks.reverse();

  for (let i = 0; i < chunks.length - 1; i++) {
    chunks[i] |= 0x80;
  }

  return chunks;
};

export const decode = (bytes) => {
  if (!Array.isArray(bytes) || bytes.length === 0) {
    throw new Error('Invalid input');
  }

  for (let i = 0; i < bytes.length; i++) {
    const byte = bytes[i];
    if (i < bytes.length - 1 && (byte & 0x80) !== 0x80) {
      throw new Error('Invalid continuation byte');
    }
    if (i === bytes.length - 1 && (byte & 0x80) !== 0) {
      throw new Error('Last byte has continuation bit set');
    }
  }

  const result = [];
  let currentValue = 0;
  for (const byte of bytes) {
    result = (result << 7) | (byte & 0x7F);
  }

  if (result > 0xFFFFFFFF) {
    throw new Error('Number exceeds 32-bit unsigned integer');
  }

  return result;
};

File: src/2025-02-23-deepclaude-try/java/exercises/practice/variable-length-quantity/src/main/java/VariableLengthQuantity.java
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;

class VariableLengthQuantity {

    List<String> encode(List<Long> numbers) {
        List<String> result = new ArrayList<>();
        for (Long number : numbers) {
            long num = number;
            List<Byte> chunks = new ArrayList<>();
            if (num == 0) {
                chunks.add((byte)0);
            } else {
                while (num != 0) {
                    byte chunk = (byte)(num & 0x7F);
                    chunks.add(chunk);
                    num >>>= 7;
                }
            }
            Collections.reverse(chunks);
            if (!chunks.isEmpty()) {
                for (int i = 0; i < chunks.size() - 1; i++) {
                    chunks.set(i, (byte)(chunks.get(i) | 0x80));
                }
                for (byte b : chunks) {
                    result.add(String.format("%02X", b & 0xFF));
                }
            }
        }
        return result;
    }

    List<String> decode(List<Long> bytes) {
        List<String> result = new ArrayList<>();
        int index = 0;
        while (index < bytes.size()) {
            long value = 0;
            int shift = 0;
            byte b;
            do {
                if (index >= bytes.size()) {
                    throw new IllegalArgumentException("Incomplete VLQ sequence");
                }
                long byteValue = bytes.get(index++);
                if (byteValue < 0 || byteValue > 255) {
                    throw new IllegalArgumentException("Invalid byte value: " + byteValue);
                }
                b = (byte) byteValue;
                long segment = (b & 0x7F);
                value |= segment << shift;
                shift += 7;
            } while ((b & 0x80) != 0 && shift <= 28);
            
            if (value > 0xFFFFFFFFL) {
                throw new IllegalArgumentException("Value exceeds 32 bits");
            }
            result.add(String.format("%08X", value));
        }
        return result;
    }
}
